### ISSUE TITLE
Update/discussion of design specification for NX 2.0 

### ISSUE THREAD

**Issue by MridulS** [networkx/networkx#1382]
As suggested in https://github.com/networkx/networkx/pull/1344 , I think we should update/further discuss the design specification accordingly. 
@hagberg @dschult @chebee7i

**Comment by ysitu** [comment]
Overall, I am against the patchwork that adds/augments `data` parameters in existing functions. It is just not maintainable. Every future change will touch a whole bunch of functions and creates headache when you try to keep the semantics of `data` consistent across the board. My preferred approach is still a new, single rule-them-all query function of which we do not need to worry about legacy.

**Comment by MridulS** [comment]
@ysitu So are you against the proposed design specification in https://github.com/networkx/networkx/wiki/Design-Specification

Towards the end of the discussion in https://github.com/networkx/networkx/pull/762 the possibility of returning an iterator for G[u] is suggested. Is there any possibility that this breaks the existing code?

**Comment by ysitu** [comment]
My opinion is that we should keep the existing interface as is (except for renaming the `_iter` functions) and start a new interface with a clean design.

As for `G[u]` behaving like an iterator, I would prefer that it behave like a set. I would like that `v in G[u]` continue to work. But changing `G[u]` alone is also a major undertaking as much code inside NetworkX itself already assumes that `G[u]` is a dict.

**Comment by chebee7i** [comment]
For some reason, my memory tells me that we had previous come to some sort of soft plan that 2.0 would try to avoid being a major wall we had to climb. We already have tons of new functionality, as is. If we focused 2.0 on renaming the `_iter` functions, then we can postpone many of the other changes for a 3.0.  3.0 doesn't have to be that far away. IMO, it is much easier for people to adapt code to backwards incompatible changes when the number of changes are small and fairly insignificant. Lumping them all into one big release might be asking for trouble. If I can find it, I will link a similar discussion that happened in NumPy. Also, we don't have to go to 2.0 right now, we could just as well do a 1.10.

**Comment by hagberg** [comment]
I remember the same discuss @chebee7i.  I agree with the approach of fixing the _iter functions and doing light work on some warts. And call that 2.0.
It could be difficult (and probably out-of-scope for a GSoC project) to redesign the entire API.  There will certainly be many opinions on what a "clean design" interface looks like.

**Comment by dschult** [comment]
Renaming the ```_iter``` functions along with what has already been committed is already a fairly large set of changes.  So I think limiting 2.0 to that is a good way forward. We don't want to wait too long between releases anyway.

**Comment by MridulS** [comment]
I made this https://github.com/networkx/networkx/wiki/NetworkX-GSoC-2015-Application:-NetworkX-2.0-API---Mridul-Seth as a rough draft. 
So I guess API change is out of the picture and 2.0 will only having renaming the _iter functions.
Will update the GSoC proposal accordingly. :)

**Comment by chebee7i** [comment]
Big changes to the API still seem like they could be part of a GSoC proposal, but the GSoC need not be tied to the release of 2.0.  IOW. 2.0 might be released with partially completed project.

**Comment by MridulS** [comment]
So should the first priority of this project be to push the changes for 2.0 or for big changes to the API?

**Comment by MridulS** [comment]
July 1, 2015 is set as the date for milestone NetworkX 2.0

**Comment by chebee7i** [comment]
That (pushing changes for 2.0) seems reasonable to me, but let others chime in too.

Regarding the timeline, we haven't stuck to those deadlines in the past---usually being overdue. I don't think it would be a huge concern if we were early either.

**Comment by ysitu** [comment]
I have changed the 2.0 deadline to after the end of GSoC. It makes sense to wait for the GSoC work.

**Comment by MridulS** [comment]
Currently `node()` returns a list, which is generated by the iterator returned by `nodes_iter()`. So we expect to do something like this in the `nodes()` function:

``` python
def nodes_iter(self, data=False):
      if data:
            return iter(self.node.items())
      return iter(self.node)     
def nodes(self, data=False):
      return list(self.nodes_iter(data=data))
```
to
``` python
def nodes(self, data=False):
      if data:
            return iter(self.node.items())
      return iter(self.node)
```
correct me if I am wrong.

**Comment by ysitu** [comment]
@MridulS Yes, we want to rename `nodes_iter` over `nodes`, etc.

**Comment by chebee7i** [comment]
Can we use `d.viewkeys()`, since we are 2.7+ now?

**Comment by hagberg** [comment]
I hope we won't use "viewkeys()" and friends.  They don't exist in Python3.

**Comment by MridulS** [comment]
Tried changing it just for nodes. Nosetests output:
``` 
Ran 2574 tests in 32.606s

FAILED (SKIP=8, errors=246, failures=116)
```
This will be fun :D

**Comment by chebee7i** [comment]
Ah right, `viewkeys in 2.7+` <--> `keys in 3.x`.

**Comment by MridulS** [comment]
We could use this http://python-future.org/reference.html?highlight=keys#future.utils.viewkeys . Making the code Python 2/3 compatible will also be a part of my project by using `future` and `six`  @Midnighter

**Comment by hagberg** [comment]
I am -1 on using future and six to put lots of Python2/3 cruft in the code.
Python 3 is here people. And it works great.

**Comment by ysitu** [comment]
@MridulS You can use ysitu/networkx@bfce79ab2af50f5a7e7769d36b8c40909985e4c9 as the starting point.

**Comment by MridulS** [comment]
@ysitu Thanks :) going through it.

**Comment by MridulS** [comment]
A rough timeline for my GSoC proposal:
##### April 27 - May 25 (Pre GSoC Period)
* Working on issues and pending PRs in https://github.com/networkx/networkx/milestones/networkx-2.0
* Discussion about the API Design specification 

#####  May 25 - June 30
* Working on renaming _iter functions.
* Updating the documentation and tests as necessary. 
* Further discussion about the API design.
* Cleaning and reorganising the current codebase as required and decided. 

##### July 1 - August 21
* Working on the implementation of the new API design.
* Cleaning and reorganising the current codebase as required and decided. 

##### August 1 -   (Post GSoC)
* Continue working on the new API design and get it ready for the next major version release. :)

@ysitu @dschult @hagberg @chebee7i Looking forward to your views. :)

**Comment by MridulS** [comment]
updated my application https://github.com/networkx/networkx/wiki/NetworkX-GSoC-2015-Application:-NetworkX-2.0-API---Mridul-Seth 
views and suggestions ?

**Comment by jtorrents** [comment]
I'm still catching up with all the discussion on 2.0 API, but I think it might be useful to put together a document that mirrors more or less Python's [PEPs](https://www.python.org/dev/peps/) . If we have a document like this it will be a lot easier to ask people that do not follow closely NetworkX development for feedback about API design. I'm thinking that would be very useful for us to have feedback from people such as Raymond Hettinger, David Eppstein, and others. It might be a lot easier for them to give feedback if we have everything explained in a single document.

Also, it is useful to look at previous discussions about a python Graph API, see:

https://wiki.python.org/moin/PythonGraphApi

I find especially useful this thread of python-dev (referenced in the above wiki page):

http://groups-beta.google.com/group/comp.lang.python/browse_thread/thread/cbca60cb36be39ed/313113af1aa077af

**Comment by ysitu** [comment]
Another item to review for the 2.0 API is the practice of importing almost everything into the top-level module.

**Comment by hagberg** [comment]
Should we make a release before 2.0?  There have been many changes and additions since the last release.  These could be used now without having to worry (much) about API changes that will be in 2.0.  
And we pushed the date of 2.0 now until September.

**Comment by dschult** [comment]
I think we have more changes already for a new release than we usually do. And presumably there will be many more between now and Sept. Maybe its a good idea to get these changes released before we start messing with API issues.

**Comment by chebee7i** [comment]
Yeah, I wouldn't mind a 1.10 release or something similar. Presumably, there will be people who will be quite slow to migrate to 2.0, so it'd be nice to get as many features out with the old API as possible.

**Comment by chebee7i** [comment]
We're gonna have a fun time updating all the PRs and issues that were tagged as 2.0. Haha. No big deal though.

**Comment by MridulS** [comment]
Should we deprecate _iter suffix in this new immediate release?

**Comment by jtorrents** [comment]
I think it's a good idea to make a release with the changes that we have now, as Dan says, we already have a lot of changes and new features.

@MridulS , I think that removing the `*_iter` functions should wait until 2.0 as it's an important change of the API, and can potentially break user code.

**Comment by ysitu** [comment]
@MridulS @jtorrents Calling out those functions as deprecated is a good idea. That paves the way for removal in 2.0.

@chebee7i I have bulk-marked all closed 2.0 Issues/PRs and a few open ones as targeting 1.10.

**Comment by MridulS** [comment]
@ysitu the current version is 1.9.1, shouldn't they be marked something like 1.9.2 ... 1.9.9

**Comment by debugger22** [comment]
@MridulS 
For `x.y.z`

`x` is increased in case of compatibility breaks.
`y` is increased in case of additional features.
`z` is increased in case of bug fixes.

`1.10` seems apt as suggested by @ysitu.

**Comment by debugger22** [comment]
There are no rules, though. Linux kernel `4.0` was released after `3.19` even though there were no compatibility breaks.

**Comment by MridulS** [comment]
@debugger22 NetworkX v1.1 is already there http://networkx.lanl.gov/archive/networkx-1.1/
https://github.com/networkx/networkx/tree/networkx-1.1

**Comment by debugger22** [comment]
This is not mathematical equality. `1.1 != 1.10` :)

**Comment by chebee7i** [comment]
Yeah, that part can be confusing, but the version number should not be interpreted as a decimal---afterall, how would one interpret 1.2.1. Periods separate integers (possibly with alpha, pre, or whatever also attached to them).

**Comment by jg-you** [comment]
:+1: on a pre-2.0 release with deprecated interfaces still available. :+1:  on `1.10` too.

**Comment by MridulS** [comment]
What timeline do you suggest regarding this release? 

As discussed earlier 2.0 should mainly have the `_iter` functions deprecated and further changes in API will be in future releases. 

PS: GSoC results out. Excited for summer :D

**Comment by ysitu** [comment]
@MridulS Yea, I am yet to get around to make a public announcement regarding GSoC. Maybe later today.

**Comment by ysitu** [comment]
@MridulS Regarding the release schedule, I have marked May 25 in the milestone. Typically we send out an RC one week in advance. So we need to have all open issues closed by May 18 or deferred till 2.0.
